java-checks/src/main/java/org/sonar/java/checks/SillyEqualsCheck.java

```diff
@@ -0,0 +1,118 @@
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+
+@Rule(
+  key = "S2159",
+  name = "Silly equality checks should not be made",
+  tags = {"bug", "unused"},
+  priority = Priority.MAJOR)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("15min")
+public class SillyEqualsCheck extends SubscriptionBaseVisitor {
+
+  private static final String MESSAGE = "Remove this call to \"equals\"; comparisons between unrelated types always return false.";
+
+  private static final MethodInvocationMatcher EQUALS_METHOD = MethodInvocationMatcher.create()
+    .name("equals")
+    .addParameter("java.lang.Object");
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree methodTree = (MethodInvocationTree) tree;
+    if (EQUALS_METHOD.matches(methodTree)) {
+      ExpressionTree firstArgument = Iterables.getOnlyElement(methodTree.arguments());
+      Type argumentType = firstArgument.symbolType().erasure();
+      Type ownerType = getMethodOwnerType(methodTree).erasure();
+      if (isLiteralNull(firstArgument)) {
+        addIssue(tree, "Remove this call to \"equals\"; comparisons against null always return false; consider using '== null' to check for nullity.");
+      } else if (ownerType.isArray()) {
+        checkWhenOwnerIsArray(tree, ownerType, argumentType);
+      } else {
+        checkWhenOwnerIsNotArray(tree, ownerType, argumentType);
+      }
+    }
+  }
+
+  private void checkWhenOwnerIsArray(Tree tree, Type ownerType, Type argumentType) {
+    if (!argumentType.isArray()) {
+      if (!argumentType.is("java.lang.Object")) {
+        addIssue(tree, "Remove this call to \"equals\"; comparisons between an array and a type always return false.");
+      }
+    } else if (!areRelated(((Type.ArrayType) ownerType).elementType(), ((Type.ArrayType) argumentType).elementType())) {
+      addIssue(tree, "Remove this call to \"equals\"; comparisons between unrelated arrays always return false.");
+    } else {
+      addIssue(tree, "Use \"Arrays.equals(array1, array2)\" or the \"==\" operator instead of using the \"Object.equals(Object obj)\" method.");
+    }
+  }
+
+  private void checkWhenOwnerIsNotArray(Tree tree, Type ownerType, Type argumentType) {
+    if (argumentType.isArray() && !ownerType.is("java.lang.Object")) {
+      addIssue(tree, "Remove this call to \"equals\"; comparisons between a type and an array always return false.");
+    } else if (argumentType.isClass() && !areRelated(ownerType, argumentType)) {
+      if (ownerType.symbol().isInterface() && argumentType.symbol().isFinal() || argumentType.symbol().isInterface() && ownerType.symbol().isFinal()) {
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Add parentheses to make the operator precedence explicit.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS864)
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Add parentheses to make the operator precedence explicit.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS864)